docs/architecture/adr-071-crypto-v2-multi-curve.md

```diff
@@ -0,0 +1,597 @@
+
+* Restructure, Refactor, and Decouple: Update the codebase to ensure modularity and future adaptability.
+
+Documentation & Community Engagement
+
+* Enhance documentation to ensure clarity, establish a good practices protocol and promote community engagement, providing a platform for feedback and collaborative growth.
+
+Backward Compatibility & Migration
+
+* Prioritize compatibility with previous version to avoid disruptions for existing users.
+
+* Design and propose a suitable migration path, ensuring transitions are as seamless as possible.
+
+
+Developer-Centric Approach
+
+* Prioritize clear, intuitive interfaces and best-practice design principles.
+
+
+Quality Assurance
+
+* Enhanced Test Coverage: Improve testing methodologies to ensure the robustness and reliability of the module.
+
+* Conduct an Audit: After implementation, perform a comprehensive audit to identify potential vulnerabilities and ensure the module's security and stability.
+
+## Technical Goals
+
+Multi-curve support:
+
+* Support for a wide range of cryptographic curves to be integrated seamlessly into the sdk in a modular way.
+
+Wide Hardware Device & Cloud-based HSM Interface Support:
+
+* Design a foundational interface for various hardware devices (Ledger, YubiKey, Thales, etc.) and cloud-based HSMs (Amazon, Azure) to support both current and future implementations.
+
+Testing:
+
+* Design an environment for testing, ensuring developers can validate integrations without compromising system integrity.
+
+New Keyring:
+
+* Design a new Keyring interface with modular backends injection system to support hardware devices and cloud-based HSMs.
+
+
+## Proposed architecture
+
+### Introduction
+
+In this section, we will first introduce the concept of a `CryptoProvider`, which serves as the main API of the new cryptographic architecture. Following this, we will present the detailed components that make up the `CryptoProvider`. Lastly, we will introduce the storage and persistence layer, providing code snippets for each component to illustrate their implementation.
+
+
+#### Crypto Provider
+
+Here we introduce the concept of `CryptoProvider`. This interface acts as a centralized controller, encapsulating the APIs for the **signing**, **verifying** and **hashing** functionalities. It acts as the main API with which the apps will interact with
+
+By abstracting the underlying cryptographic functionality, `CryptoProvider` enables a modular and extensible architecture, aka 'pluggable cryptography'. It allows users to easily switch between different cryptographic implementations without impacting the rest of the system.
+
+The `CryptoProvider` interface includes getters for essential cryptographic functionalities and metadata:
+
+
+```go
+// CryptoProvider aggregates the functionalities of signing, verifying, and hashing, and provides metadata.
+type CryptoProvider interface {
```

Comment:
This interface is still conflating client and state machine. 

The state machine itself only cares about verify. Is there a reason someone would need to implement the rest of the interface if they decide the curve is only supported over rest not cli. 

Asking people to implement these seems like it would end up in boilerplate code with panics in them. Can we avoid this

Reply:
We've through this discussion before and I think these new proposal has the same conclusions that the ones we had in that time. In the document, near the end, you'll find a section to this matter. But in short, if someone needs to run a `Verifier` over a signature means that the proper `Signer` has been written, therefore the corresponding `CryptoProvider` exists and it should be fully implemented (Verifier, Signer, Hasher, etc). So if the state machine only uses the verifier, it won't need to re-implement that code because the provider that made that signature exists and also its verifier, the state machine code would only have to instantiate that `CryptoProvider` and only use its `Verifier`.